### PR TITLE
DocumentStore and SessionProvider async and order independent

### PR DESCRIPTION
**Pull request by danielmarbach**
This change makes it possible to safely retrieve the document store or a document session without running into ordering problems. Due to the interfaces returning value tasks it is also less likely that someone accidentally tries to use the document store provider in a constructor to assign a field.

We have discussed the alternative of wrapping the document store but it turned out the synchronous nature of the document store interface would require us to come up with complicated synchronous synchronization logic until the embedded instance is started.

### CODE REVIEW

File: src/ServiceControl.Audit.Persistence.RavenDB/RavenExternalPersistenceLifecycle.cs [hunk 1]
```diff
@@ -10,41 +10,59 @@ namespace ServiceControl.Audit.Persistence.RavenDB
...
+                    Urls = [configuration.ServerConfiguration.ConnectionString],
+                    Conventions = new DocumentConventions
+                    {
+                        SaveEnumsAsIntegers = true
+                    }
+                };
+
+                if (configuration.FindClrType != null)
+                {
+                    store.Conventions.FindClrType += configuration.FindClrType;
```
Comment (mauroservienti) [inline src/ServiceControl.Audit.Persistence.RavenDB/RavenExternalPersistenceLifecycle.cs:49]: The issue is that a single configuration will start affecting independent databases and persistence seams.

**Review by mauroservienti — APPROVED** [review]

File: src/ServiceControl.Audit.Persistence.RavenDB/RavenSessionProvider.cs [hunk 2]
```diff
@@ -1,11 +1,17 @@
...
+    using System.Threading;
+    using System.Threading.Tasks;
     using Raven.Client.Documents.Session;
 
     class RavenSessionProvider(IRavenDocumentStoreProvider documentStoreProvider) : IRavenSessionProvider
     {
-        public IAsyncDocumentSession OpenSession() =>
-            documentStoreProvider.GetDocumentStore()
-                .OpenAsyncSession();
+        public async ValueTask<IAsyncDocumentSession> OpenSession(SessionOptions? options = default, CancellationToken cancellationToken = default)
```
[the remark this answers is not in the document]
Comment (danielmarbach) [reply src/ServiceControl.Audit.Persistence.RavenDB/RavenSessionProvider.cs:11]: Same as below ;)

**Comment by danielmarbach** [comment]
Maybe we should purely look at this PR from the perspective of "Does it make the current seam better and does it address the ordering problem". If the answer is yes I think we should for a moment leave the other componentization requirements out of scope.

File: src/ServiceControl.Audit.Persistence.RavenDB/RavenExternalPersistenceLifecycle.cs (hunk 1 above)
Comment (mauroservienti) [reply src/ServiceControl.Audit.Persistence.RavenDB/RavenExternalPersistenceLifecycle.cs:49]: Agreed

File: src/ServiceControl.Persistence.RavenDB/IRavenSessionProvider.cs [hunk 3]
```diff
@@ -2,10 +2,12 @@
...
 namespace ServiceControl.Persistence.RavenDB
 {
+    using System.Threading;
+    using System.Threading.Tasks;
     using Raven.Client.Documents.Session;
 
     public interface IRavenSessionProvider
     {
-        IAsyncDocumentSession OpenSession(SessionOptions? options = default);
+        ValueTask<IAsyncDocumentSession> OpenSession(SessionOptions? options = default, CancellationToken cancellationToken = default);
```
[the remark this answers is not in the document]
Comment (mauroservienti) [reply src/ServiceControl.Persistence.RavenDB/IRavenSessionProvider.cs:11]: Agreed.

**Review by bording — APPROVED** [review]

**Comment by danielmarbach** [comment]
Given that we now have https://github.com/Particular/ServiceControl/pull/4140 available we are moving forwarding with merging this.

// cc @abparticular @johnsimons @jpalac

**Pull request merged**